compiler/rustc_macros/src/diagnostics/subdiagnostic_builder.rs

```diff
@@ -0,0 +1,612 @@
+            throw_span_err!(
+                self.variant.ast().ident.span().unwrap(),
+                "subdiagnostic kind not specified"
+            );
+        };
+
+        let init = match &kind {
+            SubdiagnosticKind::Label
+            | SubdiagnosticKind::Note
+            | SubdiagnosticKind::Help
+            | SubdiagnosticKind::Warn
+            | SubdiagnosticKind::Suggestion { .. } => quote! {},
+            SubdiagnosticKind::MultiPartSuggestion { .. } => {
+                quote! { let mut suggestions = Vec::new(); }
+            }
+        };
+
+        let attr_args: TokenStream = self
+            .variant
+            .bindings()
+            .iter()
+            .filter(|binding| !binding.ast().attrs.is_empty())
+            .map(|binding| self.generate_field_attr_code(binding, &kind))
+            .collect();
+
+        let primary_span_field = self.primary_span_field.as_ref().map(|(span, _)| span);
+
+        let diag = &self.diag;
+        let name =
+            format_ident!("{}{}", if primary_span_field.is_some() { "span_" } else { "" }, kind);
+        let message = quote! { rustc_errors::fluent::#slug };
+        let call = match kind {
+            SubdiagnosticKind::Suggestion { suggestion_kind, code } => {
+                let applicability = self.applicability.take().map_or_else(
+                    || {
+                        span_err(self.span, "suggestion without `applicability`").emit();
+                        quote! { rustc_errors::Applicability::Unspecified }
+                    },
+                    |(applicability, _)| applicability,
+                );
+
+                if let Some(span) = primary_span_field {
+                    let style = suggestion_kind.to_suggestion_style();
+
+                    quote! { #diag.#name(#span, #message, #code, #applicability, #style); }
+                } else {
+                    span_err(self.span, "suggestion without `#[primary_span]` field").emit();
+                    quote! { unreachable!(); }
+                }
+            }
+            SubdiagnosticKind::MultiPartSuggestion { suggestion_kind } => {
+                let applicability = self.applicability.take().map_or_else(
+                    || {
+                        span_err(self.span, "multipart suggestion without `applicability`").emit();
+                        quote! { rustc_errors::Applicability::Unspecified }
+                    },
+                    |(applicability, _)| applicability,
+                );
+
+                if !self.has_suggestion_parts {
+                    span_err(
+                        self.span,
+                        "multipart suggestion without any `#[suggestion_part(...)]` fields",
+                    )
+                    .emit();
+                }
+
+                let style = suggestion_kind.to_suggestion_style();
+
+                quote! { #diag.#name(#message, suggestions, #applicability, #style); }
```

Comment:
`multipart_suggestion()` will fail with an assert at runtime if the list of suggestions is empty, what should be done about that? If there's only a single `#[suggestion_part(...)] spans: Vec<Span>` field, there's no way to guarantee during derive generation that it won't be empty.

Reply:
I think this is fine for now, we can always improve it later.